rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,228 @@
+
+### Automatic deployments
+
+We will enable automatic deployments for the apps listed [here](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/hieradata_aws/common.yaml#L145), but only when they meet the safety criteria of this RFC. To help with this, a rough audit of "compatibility" can be found in [Appendix A: Compatibility](rfc-128/compatibility.md).
+
+- We can immediately enable automatic deployments for apps marked with a ✅.
+- We will undertake work to address all the issues (⚠) for apps marked with a ❌.
+
+> **We will disable CD for Publishing API**. This was only enabled on a trial basis, and requires non-trivial work to bring it up to the new standard. We should avoid setting a false precedent by leaving CD enabled.
+
+The following steps must be taken as part of enabling automatic deployments:
+
+- Restrict merge access to "GOV.UK Production" [[example](https://github.com/alphagov/govuk-saas-config/pull/55)].
+
+- Tag [relevant Smoke tests](https://github.com/alphagov/smokey/blob/5dc36fc78c12838ec1c8070d6fbc535ae5433b89/docs/writing-tests.md#filtering-by-app) to run for the app [[example](https://github.com/alphagov/smokey/pull/708/commits/91a5ce94b7b96d09f5db87e63fbc3e400788de00)].
+
+- Add a note in the [Release](https://github.com/alphagov/release) app to warn about the change.
+
+- Add a PR template to warn about the change [[example](https://github.com/alphagov/smart-answers/blob/3509ba3e68c62a206874c96111628dd65adae689/.github/pull_request_template.md)].
+
+  - Backfill the PR template as a comment on existing PRs.
+
+- Enable automatic deployments in a Puppet PR [[example](https://github.com/alphagov/govuk-puppet/pull/10590)].
+
+  - Comment on how the app meets the criteria of this RFC.
+
+> **We can be flexible with coverage**. We can defer adding internal repo tests if the code has no impact on the running app e.g. rake tasks [[1](https://github.com/alphagov/email-alert-api/blob/59fc71a58317ef2998f2c0ef102020da3ca9df96/lib/tasks/support.rake)]. But we should not defer removing any unused code.
+
+### Slack notifications
+
+We need to ensure we are notified about a failure in the automation, so that the automation is no worse than the current, manual process. This has already been done for the `Deploy_App_Downstream` job [[1](https://github.com/alphagov/govuk-puppet/pull/10532)] [[2](https://github.com/alphagov/govuk-puppet/pull/10601)].
+
+So that we cover the entire deployment pipeline, we will also update [the `Deploy_App` job](https://github.com/alphagov/govuk-puppet/blob/f57335f6027487b5b2d2205a35ceb411776f09ce/modules/govuk_jenkins/templates/jobs/deploy_app.yaml.erb) to post a Slack message in the #govuk-developers Slack channel, whenever it fails due to an unexpected error.
+
+### Delete [publishing-e2e-tests](https://github.com/alphagov/publishing-e2e-tests)
```

Comment:
Do the e2e tests provide any feature coverage which doesn't overlap with Smokey?

Reply:
Yes, the e2e tests focus on actions that have side effects - such as altering content. Therefore they're best suited to an isolated environment where flakiness can't be caused by other users (and one of the pain points is the resources needed to build and configure a mini GOV.UK). The Smokey test suite does (nearly) idempotent actions that don't produce noticeable side effects, which make them safe to run against production environment.

A closer comparison to where coverage would be lost would be to compare whether there are things in e2e tests that aren't covered in test suites of individual apps. We'd very much hope that isn't the case as e2e tests weren't intended to be a replacement or alternative to testing in individual apps.

Reply:
I completely agree that system testing (e2e tests) shouldn't be overused and that most tests should be at the bottom of the pyramid (lots of unit tests > some integration tests > few e2e tests).

On the other hand it seems to me entirely unrealistic to imagine that a system of 50+ components doesn't need any system testing at all and can be tested entirely through unit and integration testing. If there are some superfluous tests in the e2e test suite which should be deleted then that's one thing, but to delete the whole lot sounds like it would leave some [serious gaps](https://gfycat.com/happygoluckyseriouschimneyswift).

Reply:
The question did come up before: should we try to delete the E2E tests in this RFC?

One of the problems I had when writing the safety criteria is: where do the E2E tests fit in? I don't think we should omit them altogether, since they are clearly relevant, at least in name, to the idea of "safety". To avoid leaving them in limbo, I had to consider what value they actually add on top of the tests we have already.

I concluded they add very little, if any, value:

- We've avoided adding new tests to them e.g. notice the absence of Content Publisher.
- We don't seem to be testing key workflows with them e.g. [these tests are oddly specific](https://github.com/alphagov/publishing-e2e-tests/tree/master/spec/collections_publisher).
- They completely omit our email and support systems (nothing about these is tested).
- They have a very high maintenance cost, not least from adding 5+ mins to every PR build.

Really, it comes down to this: we have a poor system for E2E tests that we actively avoid; we're not going to invest in it, and leaving it as-is sets a bad precedent for continuing to avoid E2E tests in future. We've always been scared to delete them in a standalone change (no one seems to know why). The fact that none of them are a safety criteria for CD is possibly the best indicator we'll ever have that they are, in fact, useless.

☝️ Some strong emotions here (I've feel I've wasted precious life on these things). I think it's worth pointing out that I did make a real effort to go [the other way](https://github.com/alphagov/govuk-rfcs/blob/continuous-deployment/rfc-128-continuous-deployment.md#check-detailed-features-of-govuk-work), but I came up blank. Despite all the words, there's a simple way we can make progress: can you (or anyone) point out something critical in the E2E tests that isn't covered by the other tests discussed here? If we can do that, then perhaps there is an argument for keeping them.

Reply:
I haven't had to run the end-to-end tests too much manually during my time on GOV.UK, so there's a risk that I might not be speaking from the voice of experience here, but I broadly agree with @sengi above. There may be tests that aren't in a critical path in the end-to-end tests, which we perhaps don't care about from a CD perspective. On the other hand, there may also be tests in the end-to-end tests which, even if they aren't critical, aren't replicated elsewhere and would form a gap were they to be deleted.

The reasoned approach (in my opinion) would be to go ahead with CD without including the end-to-end tests as part of that, but _not_ deleting them at the same time. Let's minimise the moving parts here.

At the same time, if we strongly feel that the end-to-end tests serve no purpose then let's open an RFC, even if it's purely to re-iterate the comments that have already been made in this RFC.

Reply:
Thanks for your thoughts @karlbaker02. I've been talking to @sengi about this and both agree that the current argument for deleting the E2E tests here doesn't hold.

However, I still wish to pursue this as part of this RFC, because it's a natural consequence of totally ignoring these tests when considering the safety of the changes we make to our apps. 

I did originally consider a separate RFC, but was put off by previous discussions around this: people really don't like them, but they also seem to be scared of deleting them. This RFC presents a great opportunity, since it brings both of these issues together in a constructive way. Conversely, I can easily imagine an RFC about deleting the end tests going the same way as discussions in the past: "we don't like them, we don't want to delete them, but we don't have time to improve them, either". I feel very strongly that we should do _something_ here.

I've talked to @sengi about a few different options, besides deletion. We eventually agreed it's worthwhile to continue arguing for deletion, but with some different words. @sengi what do you think:

### Delete [publishing-e2e-tests](https://github.com/alphagov/publishing-e2e-tests)

In establishing the safety criteria for automatic deployments, we had to consider these tests. As explained in the rejected approaches, we chose to avoid them entirely. This brings their existence into question.

We believe these tests add little or no value to GOV.UK. They significantly reduce our agility by adding 5+ minutes onto every PR build, to the extent that we sometimes remove them as a required check. They have very significant maintenance issues and only ever break for reasons unrelated to the features they are testing. Despite these signficiant problems, we invest very little effort in improving them, and actively avoid extending them.

We believe what we gain by deleting them far outweighs the high cost of keeping them.

Reply:
Sounds reasonable to me.

Reply:
👍 - I've added a commit for this: https://github.com/alphagov/govuk-rfcs/pull/128/commits/6d2dbd30282de9f7b626f15c99425df88ebc83c4

Reply:
Thanks Ben. If you're both happy then that's great, and it makes sense about bringing the issue together with this RFC on CD. Happy to go with consensus 👍

Reply:
I've given this some thought too, and I'm persuaded that removing these tests once we've improved coverage in the other test suites is the right thing to do.

1) This isn't us removing _all_ of our end-to-end tests, as we still have smokey (albeit that smokey doesn't aim to cover state-changing bits of publishing)

2) The publishing-e2e tests always felt like they were running too early in the pipeline (on pull requests). We should have enough confidence from unit / integration / contract testing that a particular change is safe to merge. Perhaps there's a spot for some more end-to-end testing later in the pipeline, but the publishing-e2e test probably aren't the right fit for that job (e.g. they don't run against realistic infrastructure).

I can see the argument that the deletion of these tests should be discussed on a separate RFC - it would have been nice to take the time to quantify exactly how much pain they're causing us at the moment, and compare that to some analysis of how many bugs they catch. But doing that analysis would be a significant amount of work, and I'm pretty confident it wouldn't change our decision. Not worth it. Still, I can't shake the feeling that it's being snuck through as a [Rider](https://en.wikipedia.org/wiki/Rider_(legislation)) to get a controversial decision made on the back of an uncontroversial one. If that's deliberate, then congratulations, the strategy has worked on me 😆.

@bevanloon - what do you think?

Reply:
@richardTowers @bevanloon I've made another adjustment to this section, to help make sure it's achievable in practice: https://github.com/alphagov/govuk-rfcs/pull/128/commits/7d6d8de0c346bec2d05a7d9ca18bccf1d0563aef.

Reply:
I've had a long chat with Richard about this and I'm happy to let the publishing e2e tests go once we've done the work to bring our non-frontend apps up to a continuous deployment state as suggested.

Reply:
Thanks @bevanloon! And everyone for taking the time to think through the decision - it's by no means any easy call to make. Following our discussion, I've added a commit to adjust the wording here: https://github.com/alphagov/govuk-rfcs/pull/128/commits/3402d4cd32b6e9dd383005292a58414e81c481a9.

I'll resolve this now, but anyone can comment if they think there's more to discuss.